tooling/v8-snapshot/src/doctor/determine-deferred.ts

```diff
@@ -14,49 +13,65 @@ export async function determineDeferred (
   cacheDir: string,
   opts: {
     nodeModulesOnly: boolean
-    previousDeferred: Set<string>
-    previousHealthy: Set<string>
-    previousNoRewrite: Set<string>
     forceNoRewrite: Set<string>
-    useHashBasedCache: boolean
     nodeEnv: string
     integrityCheckSource: string | undefined
   },
 ) {
   const jsonPath = path.join(cacheDir, 'snapshot-meta.json')
+  const usePreviousSnapshotMetadata = (!process.env.V8_SNAPSHOT_FROM_SCRATCH || !['1', 'true'].includes(process.env.V8_SNAPSHOT_FROM_SCRATCH)) && await canAccess(jsonPath)
+  const { deferredHash, norewrite, deferred, healthy } = usePreviousSnapshotMetadata ? require(jsonPath) : { deferredHash: '', norewrite: [], deferred: [], healthy: [] }
+  const hashFilePath = await findHashFile(projectBaseDir)
+  const currentHash = await createHashForFile(hashFilePath)
+  const res = await matchFileHash(hashFilePath, deferredHash)
 
-  let hashFilePath: string | undefined
-  let hash
+  let nodeModulesHealthy: string[] = []
+  let projectHealthy: string[] = []
+  let currentHealthy = opts.nodeModulesOnly ? nodeModulesHealthy : healthy
 
-  if (opts.useHashBasedCache) {
-    hashFilePath = await findHashFile(projectBaseDir)
-    assert(
-      hashFilePath != null,
-      `Unable to find hash file inside ${projectBaseDir}`,
-    )
+  healthy.forEach((dependency) => {
```

Comment:
Going to admit I don't fully grok the finer details of some parts of this PR (like this). I get the general idea though, and it looks like this makes sense.